lib/src/Base/Algo/ClenshawCurtis.cxx

```diff
@@ -0,0 +1,212 @@
+{
+  const UnsignedInteger inputDimension = discretization_.getSize();
+  if (interval.getDimension() != inputDimension) throw InvalidArgumentException(HERE) << "Error: expected an interval of dimension=" << inputDimension << ", got dimension=" << interval.getDimension();
+  if (function.getInputDimension() != inputDimension) throw InvalidArgumentException(HERE) << "Error: expected a function of input dimension=" << inputDimension << ", got input dimension=" << function.getInputDimension();
+  // Compute the volume of the interval
+  const Scalar volume = interval.getVolume();
+  Point integral(function.getOutputDimension());
+  if (volume == 0.0) return integral;
+  // Adapt the nodes to the bounds of the interval
+  adaptedNodes = nodes_ * (interval.getUpperBound() - interval.getLowerBound())/2. + interval.getLowerBound()+(interval.getUpperBound() - interval.getLowerBound())/2.;
+  // Compute the function over the adapted nodes
+  const Sample values(function(adaptedNodes));
+  // Compute the integral
+  for (UnsignedInteger i = 0; i < values.getSize(); ++i)
+    integral += values[i] * weights_[i];
+  return std::pow(0.5,inputDimension)*integral * volume;
+}
+
+/* Generate nodes and weights */
+void ClenshawCurtis::generateNodesAndWeights()
+{
+  // First, generate the 1D marginal rules over [0, 1]
+  const UnsignedInteger dimension = discretization_.getSize();
+  if (dimension == 0) throw InvalidArgumentException(HERE) << "Error: expected a positive dimension";
+  Collection<Point> marginalNodes(dimension);
+  Collection<Point> marginalWeights(dimension);
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    const UnsignedInteger integrationNodesNumber(discretization_[i]);
+    if (integrationNodesNumber == 0) throw InvalidArgumentException(HERE) << "Error: the discretization must be positive, here discretization[" << i << "] is null.";
+    // Check if we already computed this 1D rule
+    // We use the value 'dimension' as a guard
+    UnsignedInteger indexAlreadyComputed = dimension;
+    for (UnsignedInteger j = 0; j < i; ++j)
+      if (discretization_[j] == integrationNodesNumber)
+      {
+        indexAlreadyComputed = j;
+        break;
+      }
+    // If indexAlreadyComputed < dimension we found a match
+    if (indexAlreadyComputed < dimension)
+    {
+      marginalNodes[i] = marginalNodes[indexAlreadyComputed];
+      marginalWeights[i] = marginalWeights[indexAlreadyComputed];
+    } // A match found
+    else
+    {
+      marginalNodes[i] = Point(integrationNodesNumber);
+      marginalWeights[i] = Point(integrationNodesNumber);
+      for (UnsignedInteger k = 0; k < integrationNodesNumber; ++k)
+      {
+        const Scalar theta_k = k*M_PI/(integrationNodesNumber-1);
+        Scalar ck = 0;
+        if (k == 0 || k == integrationNodesNumber-1)
+        {
+        ck = 1.;
+        }
+        else
+        {
+        ck = 2.;
+        }
+        // Nodes
+        marginalNodes[i][k] = cos(theta_k);
```

Comment:
a small detail, it's better to write:
`marginalNodes(i, k) = cos(theta_k);`
to avoid some copies